### PR TITLE
Make entities use IslandoraSolrQueryProcessor - the sequel

### PR DESCRIPTION
**Pull request by whikloj**
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-1594

# What does this Pull Request do?

Change the Islandora_solr_entities to use the IslandoraSolrQueryProcessor to build queries.

# What's new?
Build and execute queries using the IslandoraSolrQueryProcessor and therefore apply filters and namespace restrictions.

Also @rosiel added some fixes to account for IslandoraSolrQueryProcessor stripping +'s

# How should this be tested?

This should not have a large effect unless you use namespace restrictions in which case entities from other namespaces should not be accessible.

# Additional Notes:
I apologize that I don't really use this module and @rosiel while "reviewing" my initial branch ended up fixing (hopefully) most of the problems. So if there is someone that uses entities that could have a look that would be great.

Example:
* Does this change require documentation to be updated? no
* Does this change add any new dependencies? no
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? no
* Could this change impact execution of existing code? no

# Interested parties
@Islandora/7-x-1-x-committers, @nhart (opening the ticket), @bryjbrown (as an avid scholar users), @rosiel

### CODE REVIEW

**Comment by whikloj** [comment]
Ok I think I got this back to a cleaner state. My apologizes for the trouble.

**Comment by DiegoPino** [comment]
@whikloj no worries. Was more a `git` glitch than a human one. Looked like that rebase got done in whatever order git thought it was best.  Hope this gets into 7.x before code freeze. thanks

**Comment by rosiel** [comment]
OK good I'm so sorry about the rebase nightmare. I've done code in this so i can't review/commit... does anyone have time?

**Comment by rosiel** [comment]
ping @Islandora/7-x-1-x-committers  does anyone have time to review/merge before code freeze? :D

**Comment by DiegoPino** [comment]
@rosiel I can really just check code for consistency (time), but won't be able to make a test case in vagrant, create objects, etc. So hopefully someone else can truly test. I leave the decision to you. if feel that is enough (probability of adding bugs is not too high) ping me in IRC.

File: includes/entities_rss.inc
Comment (rosiel) [inline includes/entities_rss.inc]: @whikloj could you swap those + signs for %2B please (no extra spaces!) or the query picks up stuff it shouldn't. Sorry I missed this in the first pass.

**Review by rosiel — CHANGES_REQUESTED** [review]
I pulled and tested this on a fresh box, for the six functionalities listed in the Jira ticket.  Once the + signs are fixed on the RSS feed, I'm pretty confident in this and I'd give my approval for merging.

**Comment by whikloj** [comment]
@rosiel I dumped the **+** signs and just added an `AND`, that seems to be what you want right?
```
RELS_EXT_hasModel_uri_mt:\"$content_model\" AND mods_name_personal_displayForm_mt:\"($identifier)\"
```
A specific content-model _and_ a specific entity id. It seemed to work for my simple test, but I'll wait for your approval.

**Review by rosiel — APPROVED** [review]

**Comment by DiegoPino** [comment]
@whikloj @rosiel sadly I had to choose between pulls to make it to the code freeze since both were touching the same piece of code and I had to bend everything a bit because of Travis-CI. This was reported as a bug, so I went for https://github.com/Islandora/islandora_solution_pack_entities/pull/103. 
Will need sadly a rebase or cherry-pick and 2 pulls to make it into the release, but we have some extra time to test. Thanks again.

**Comment by DiegoPino** [comment]
@rosiel look at https://github.com/Islandora/islandora_solution_pack_entities/pull/112/conflicts . Seems like the conflict is not much, but probably there could be more change needed? not sure

**Comment by whikloj** [comment]
I can rebase, I hope

File: includes/callbacks.inc
Comment (qadan) [inline includes/callbacks.inc]: Just as a note: includes/query_processor.inc is included in the files[] declarations in islandora_solr.info, so it'd be okay to cut out these module_load_include()s

File: islandora_entities.module
Comment (qadan) [inline islandora_entities.module]: Could we use a function here that explicitly returns a boolean so we can be clear what we're looking for here, e.g., `!empty((string) $identifier)`? I'm assuming we're just making sure it's not an empty string, but it'd be nice to be clear due to PHP truthiness nonsense

File: includes/utilities.inc
Comment (qadan) [inline includes/utilities.inc]: Same deal here as below on islandora_entities.module:615; I dunno why I did this from the bottom up. Something to do with being left-handed, I'm sure (those devil-handers)

File: includes/utilities.inc
Comment (qadan) [inline includes/utilities.inc]: Could we grab the variable once here before doing the check, and possibly use `!empty()` instead of doing a direct comparison on string length?

File: includes/utilities.inc
Comment (qadan) [inline includes/utilities.inc]: Same thing as below on includes/utilities.inc:219; dammit I did it again :/

File: includes/utilities.inc
Comment (qadan) [inline includes/utilities.inc:231]: If we `return array()` here, we can skip the `if ($search_results)` check below.

File: includes/utilities.inc
Comment (qadan) [inline includes/utilities.inc:149]: It actually might be more prudent here to just `return NULL`, since if we don't have results we don't need to do any of the logic below. That way, we could also cut out the `if ($results)` check, since `$results` is always going to be an array with >= 0 items.

**Review by qadan — COMMENTED** [review]
I did a quick code review here; all the functional stuff seems to exactly mirror previous functionality just fine (even though on closer inspection some of that previous functionality appears to be broken and might benefit from a separate ticket to get it in order ...). I do have a couple of code cleanup items that can be applied though.

**Comment by whikloj** [comment]
Sorry @qadan just saw your comments, I will update.

File: includes/utilities.inc
Comment (whikloj) [reply includes/utilities.inc:149]: This is only if you have an Exception, if your search returns 0 results we also want to `return NULL`. I can clean this up some more though.

Comment (whikloj) [reply includes/utilities.inc:149]: Sorry that note was _not_ clear. I mean.... I can `return NULL` in the Exception, but I still need the `if ($results)` check in case the search returns 0 results. So I am saving time in the case that the user is going to get an Exception anyways. 

I'm not super against do a `return` here, but I feel like we normally don't in Exceptions. But maybe I'm thinking of old code?

File: includes/utilities.inc
Comment (whikloj) [reply includes/utilities.inc:231]: Same as above, I can `return array()` but the check is also for 0 results.

File: includes/utilities.inc
Comment (qadan) [reply includes/utilities.inc:149]: I think we should be fine in the case of 0 results; `$qp->islandoraSolrResult['response']['objects']` should be an empty array if no results are found, so it should be fine to pass along to `foreach`.

I'm not really for/against doing a `return NULL` up there. I know we do it in a few other places, but I don't think we really have a 'best practices' for this case; I'm just pointing it out for the sake of pointing things out while sporting a fancy code review hat

Comment (qadan) [reply includes/utilities.inc:149]: @adam-vessey is yelling at me that the objects array won't be set in the case where an exception is thrown, so ignore that haha

File: includes/callbacks.inc
Comment (qadan) [inline includes/callbacks.inc:89]: One other thing (haha sorry 😅) - the `fl` param isn't used in `internalSolrParams` and won't be set by `buildQuery()` - and by extension, `buildAndExecuteQuery()`. So in cases where an `fl` is required here, it's going to be missed unless we `buildQuery()`, explicitly set `solrParams['fl']`, and then `executeQuery()`. This happens in a few other places throughout here.

Comment (whikloj) [reply includes/callbacks.inc:89]: Probably doesn't matter, as the entire Solr record is included under `solr_doc` I can remove these. But as they don't affect functionality, I'd rather leave them. I think a subsequent PR to clean up bad function signatures, do a full module code sniff, etc is in order.

Comment (qadan) [reply includes/callbacks.inc:89]: Fair point; I'm ignoring the overall weirdness here carrying over from previous functionality so we might consider some old fashioned module cleanup to be part of a new ticket

**Comment by whikloj** [comment]
Sorry for the delay folks, but after about a month. I think this is ready. 

Also, I'll note that this PR seems to solve https://jira.duraspace.org/browse/ISLANDORA-1949

Tagging @bondjimbond for ISLANDORA-1949

**Comment by whikloj** [comment]
_bump_

**Comment by DiegoPino** [comment]
@whikloj i will wait for @bondjimbond a bit more. This week is release week. So not much in hurry with this. But If any other @Islandora/7-x-1-x-committers wants to review this or already tested and wants to merge, please go ahead.

**Comment by bondjimbond** [comment]
Not ready for merge yet..

Expected behaviour: Persons in the same namespace will display one another in the "Other scholars in department" box, while persons in other namespaces will not display.

Actual behaviour: the "Other scholars in department" box disappears entirely, even when Persons share the same namespace.

**Comment by bondjimbond** [comment]
Original:
screen shot 2017-04-24 at 11 10 31 am

This PR:
screen shot 2017-04-24 at 11 10 52 am

**Comment by rosiel** [comment]
@bondjimbond I am amused by your test objects! :) But I'm not able to replicate the results. When you set up the namespace restrictions, did you do the same things under solr query default AND Islandora configuration?  (Also, does your original have one less citation visible than the Pull Request, or was it just a result of screenshot cropping?)

FWIW, here are my results, with islandora: and ir: namespaces enabled (solr + islandora) and the rest (including secret:) blocked: 

| Object        | Pre-PR           | Post-PR  |
| ------------- |:-------------:| -----:|
| islandora:6 (fellow scholar)  | displayed | displayed|
| secret:1 (fellow scholar)  |   hidden   |   hidden |
| ir:1 (citation) |   displayed  |    displayed |
| secret:3 (citation) |displayed | hidden|

**Comment by bondjimbond** [comment]
This is without setting namespace restrictions -- these objects are all in the same site.  (RE the citation list: just cropping.)

So here are my Persons:

islandora:57 (Darth Vader)
islandora:56 (Emperor Palpatine)
rebels:2 (Luke Skywalker)

Each is assigned to the Galactic Empire department.

On 7.x branch, everyone appears in one another's Other Scholars in Galactic Empire, regardless of namespace.

On 7.x-ISLANDORA-1594-take-2, the Other Scholars block is just gone.

**Comment by bondjimbond** [comment]
I don't think it's a configuration thing, either; I've scanned the config pages for both versions, and they're identical. I'm only using the defaults.

**Comment by whikloj** [comment]
I'll give this a go tonight, I was not aware of that box so I never tested for it.

**Comment by whikloj** [comment]
Ok, so this is what I am seeing. Let me know what you expect.

Without namespace restrictions I can add multiple scholars to a department (with different namespaces) and see them listed on each other's page.

@bondjimbond can I ask you to check that your entities are indexed correctly into Solr. Not sure why they wouldn't but it would explain it if the Emporer did not get indexed.

**Comment by bondjimbond** [comment]
They're correctly indexed for sure, unless changing Entities branches somehow affects the indexing. 

I'll try again on a fresh box, though, and see what I get..

**Comment by bondjimbond** [comment]
OK -- don't know what went weird in my Vagrant box with your branch, but in a fresh install of the 7.x-1.9 dev box, this PR works well.

**Comment by DiegoPino** [comment]
@bondjimbond @whikloj @rosiel i see agreement here 😄 . Will wait a bit more in case @rosiel wants to add a comment and will merge after that. Thanks!

**Comment by qadan** [comment]
Re-opening [ISLANDORA-1594](http://jira.duraspace.org/browse/ISLANDORA-1594); there does indeed appear to be an issue here w/r/t the [comment up here](https://github.com/Islandora/islandora_solution_pack_entities/pull/112#discussion-diff-103994917R91) about the Solr results being altered that would apply to a standard use case: if the site has the `islandora_solr_limit_result_fields` option enabled and the relevant search field is not in the list of `result_fields` fields, `prepareSolrDoc()` is going to wipe those fields out, and no results are ever going to be passed onto the autocomplete.

As a side note, the `fl` wouldn't really work anyway; `fl` takes a comma-separated string and passes it straight on to Solr, as opposed to an array. I think limiting the results fields is still a good idea (Solr doc could get huge just for an autocomplete); just would have to be broken out into a structure like:

```php
 $qp->buildQuery($query, $params);
 $qp->solrParams['fl'] = "{$mappings[$type]['title']},PID";
 $qp->executeQuery(FALSE);
```

Otherwise, at the very least, `$alter_results` needs to be `FALSE` for `executeQuery()`.

**Pull request merged**
